### PR TITLE
enhancement: 2FA mobile app support

### PR DESCRIPTION
**Pull request by chilek**
Works for me ;-)

### CODE REVIEW

**Comment by chilek** [comment]
CC: https://github.com/jwilk/mbank-cli/issues/72

File: mbank-cli [hunk 1]
```diff
@@ -1213,25 +1213,53 @@ sub do_login  ## no critic (ExcessComplexity)
         $error_code eq ''
             or scraping_error("login.2fa.init.error: $error_code");
         my $auth_mode = $data->{AuthMode} // '';
+        my $content = '';
+        my $tran_id = '';
         if ($auth_mode eq 'NAM') {
-            known_bug(72, '2FA via mobile app is not supported yet');
+            print "Waiting for confirm click in mobile app...\n";
```
Comment (jwilk) [inline mbank-cli]: All diagonostics must go to stderr, not stdout.

File: mbank-cli [hunk 2]
```diff
@@ -1213,25 +1213,53 @@ sub do_login  ## no critic (ExcessComplexity)
...
+                    "$root_url/api/auth/status",
+                    @auth_headers,
+                    'data-type' => 'json',
+                    'Content' => encode_json({
+                        'TranId' => $tran_id
+                    })
+                );
+                $doc = download($request);
+                $data = decode_json($doc->{content})
+                    // scraping_error("login.2fa.exec: $EVAL_ERROR");
```
Comment (jwilk) [inline mbank-cli]: Too much copy&pasting, eh? :-P

File: mbank-cli [hunk 3]
```diff
@@ -1213,25 +1213,53 @@ sub do_login  ## no critic (ExcessComplexity)
...
+                $error_code = $data->{ErrorCode} // '?';
+                $error_code eq ''
+                    or scraping_error("login.2fa.exec.error: $error_code");
+                if ($data->{Status} eq 'Prepared' || $data->{Status} eq 'PreAuthorized') {
+                    sleep(1);
+                }
+            } while ($data->{Status} eq 'Prepared' || $data->{Status} eq 'PreAuthorized');
```
Comment (jwilk) [inline mbank-cli]: That's too too repetitive. :-P

File: mbank-cli [hunk 4]
```diff
@@ -1213,25 +1213,53 @@ sub do_login  ## no critic (ExcessComplexity)
...
+                    })
+                );
+                $doc = download($request);
+                $data = decode_json($doc->{content})
+                    // scraping_error("login.2fa.exec: $EVAL_ERROR");
+                $error_code = $data->{ErrorCode} // '?';
+                $error_code eq ''
+                    or scraping_error("login.2fa.exec.error: $error_code");
+                if ($data->{Status} eq 'Prepared' || $data->{Status} eq 'PreAuthorized') {
+                    sleep(1);
```
Comment (jwilk) [inline mbank-cli]: PerlCritic complains that the return value from `sleep()` is not checked.
I don't believe that `sleep()` can actually fail on any sane system, but hey, better safe than sorry.

File: mbank-cli (hunk 1 above)
Comment (chilek) [reply mbank-cli]: That's informational message not diagnostic. In my wrapper script I assume that if stderr is not empty some kind of problem has appeared, but wating for confirm click is not a problem!

File: mbank-cli (hunk 3 above)
Comment (chilek) [reply mbank-cli]: Fixed before the moment.

File: mbank-cli (hunk 1 above)
Comment (jwilk) [reply mbank-cli]: I can see your point; but this breaks machine-readabilty of stdout, which is more important than machine-readability of stderr, IMO.

> Enter sms password prompt is not diagnostic too.

SMS passwd prompt is actually printed on tty:
```
$ ./mbank-cli debug-sms-password 2019-09-16 1 >/dev/null 2>&1
SMS password (2019-09-16, #1):
```
I guess we could do the same for this message. Would that would for you? And what do we do when there's no tty available?

**Comment by chilek** [comment]
PR#75 - merge please.

**Comment by jwilk** [comment]
Is `pu/72` good to go (instead of this PR)?

**Comment by chilek** [comment]
Yes I confirm as I use it on my production environment ;-)

**Comment by jwilk** [comment]
Yay! Committed as 667938ad185e99eea8c9b7f8af91e693dfefd6e0.

**Comment by jwilk** [comment]
BTW, feel free to add yourself to the list of developers at `doc/AUTHORS` and to the list of copyright holders at the top of `mbank-cli`.

**Comment by chilek** [comment]
Btw. have a look of trusted device approach in twofactor_gauthentication roundcube plugin especially at its key method:
https://github.com/alexandregz/twofactor_gauthenticator/blob/master/twofactor_gauthenticator.php#L440-L485
As we can see they use only user-agent here, but others can take more web browser properties into account :(

**Comment by chilek** [comment]
> BTW, feel free to add yourself to the list of developers at doc/AUTHORS and to the list of copyright holders at the top of mbank-cli.

PR#77

**Pull request closed without merging**